### PR TITLE
Add CVE-2021-3564

### PR DESCRIPTION
**Pull request by MenV1S**
Попробовал добавить свой детект для уязвимости. Протестировал на версиях 3.12.9 (еще не появилась), 3.13 (уязвима), 4.14.194(уязвима), 4.14.236 (исправлена)

### CODE REVIEW

**Comment by evdenis** [comment]
Commit message says that this is CVE-2017-3564, but the filename describes it as CVE-2021-3564.

File: cvehound/cve/CVE-2021-3564.cocci [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/// Files: net/bluetooth/hci_core.c
+/// Fixes: 6a137caec23aeb9e036cdfd8a46dd8a366460e5d
```
Comment (evdenis) [inline cvehound/cve/CVE-2021-3564.cocci:2]:
```
/// Fix: 6a137caec23aeb9e036cdfd8a46dd8a366460e5d
/// Detect-To: b78752cc71d86998d3b77d873c61d6ffdb7a2142
```

Comment (evdenis) [reply cvehound/cve/CVE-2021-3564.cocci:2]: b78752cc71d86998d3b77d873c61d6ffdb7a2142 seems like the commit where the bug appeared the first time

File: cvehound/cve/CVE-2021-3564.cocci [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+
+
+static int hci_dev_do_open(struct hci_dev *hdev)
+{
+	...
+	flush_work@p_cmd_flush(&hdev->cmd_work);
+	flush_work@p_rx_flush(&hdev->rx_work);
+	...
+}
```
Comment (evdenis) [inline cvehound/cve/CVE-2021-3564.cocci:20]: I would write it like this
```
@err exists@
identifier hdev;
position p;
@@

\(hci_dev_open\|hci_dev_do_open\)(...)
{
        ...
*       \(flush_work@p\|tasklet_kill@p\)(&hdev->\(cmd_work\|cmd_task\));
        \(flush_work\|tasklet_kill\)(&hdev->\(rx_work\|rx_task\));
        ...
}
```

File: cvehound/cve/CVE-2021-3564.cocci [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+
+@script:python depends on detect@
+p << err.p_cmd_flush;
```
Comment (evdenis) [inline cvehound/cve/CVE-2021-3564.cocci:24]: p << err.p;

File: cvehound/cve/CVE-2021-3564.cocci [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+@@
+
+coccilib.report.print_report(p[0], 'ERROR: CVE-2021-3564')
```
Comment (evdenis) [inline cvehound/cve/CVE-2021-3564.cocci:27]: because there is already p[0] here

**Review by evdenis — CHANGES_REQUESTED** [review]
In the the rule should like these:
```
/// Files: net/bluetooth/hci_core.c
/// Fix: 6a137caec23aeb9e036cdfd8a46dd8a366460e5d
/// Detect-To: b78752cc71d86998d3b77d873c61d6ffdb7a2142

virtual detect

@err exists@
identifier hdev;
position p;
@@

\(hci_dev_open\|hci_dev_do_open\)(...)
{
        ...
*       \(flush_work@p\|tasklet_kill@p\)(&hdev->\(cmd_work\|cmd_task\));
        \(flush_work\|tasklet_kill\)(&hdev->\(rx_work\|rx_task\));
        ...
}


@script:python depends on detect@
p << err.p;
@@

coccilib.report.print_report(p[0], 'ERROR: CVE-2021-3564')
```

And pass the tests with:
```
$ pytest --cve CVE-2021-3564
$ pytest --cve CVE-2021-3564 --runslow
```

File: cvehound/cve/CVE-2021-3564.cocci (hunk 1 above)
Comment (MenV1S) [reply cvehound/cve/CVE-2021-3564.cocci:2]: If we added tasklet_call function into the rule shouldn't we change 'Detect-To' field?

Comment (evdenis) [reply cvehound/cve/CVE-2021-3564.cocci:2]: Commit b78752cc71d86998d3b77d873c61d6ffdb7a2142 contains tasklet_kill() function call.

**Comment by evdenis** [comment]
Please, push the changes and I'll merge the rule.

**Comment by evdenis** [comment]
Merged, thanks! https://github.com/evdenis/cvehound/commit/68822be41e7984e9212a890242d353edcdc769e5

**Pull request closed without merging**